lib/app/endpoint/terminology.rb

```diff
@@ -0,0 +1,224 @@
+            ]
+          )
+          capability.format = ['xml', 'json']
+        end
+        respond_with_type(capability, request.accept, 200)
+      end
+
+      get '/ValueSet/?:id_param?/$validate-code', provides: ['application/fhir+json', 'application/fhir+xml'] do
+        valueset_validates_code
+      rescue StandardError => e
+        issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'structure', details: { text: e.message })
+        return respond_with_type(FHIR::OperationOutcome.new(issue: issue), request.accept, 400)
+      end
+
+      post '/ValueSet/?:id_param?/$validate-code', provides: ['application/fhir+json', 'application/fhir+xml'] do
+        body = request.body.read
+        parsed_body = FHIR.from_contents(body)
+        valueset_validates_code(parsed_body, params[:id_param])
+      rescue StandardError => e
+        issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'structure', details: { text: e.message })
+        return respond_with_type(FHIR::OperationOutcome.new(issue: issue), request.accept, 400)
+      end
+
+      get '/CodeSystem/?:id_param?/$validate-code', provides: ['application/fhir+json', 'application/fhir+xml'] do
+        codesystem_validates_code(nil, parameters[:id_param])
+      end
+
+      post '/CodeSystem/?:id_param?/$validate-code', provides: ['application/fhir+json', 'application/fhir+xml'] do
+        body = request.body.read
+        parsed_body = FHIR.from_contents(body)
+        codesystem_validates_code(parsed_body, params[:id_param])
+      rescue StandardError => e
+        issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'structure', details: { text: e.message })
+        return respond_with_type(FHIR::OperationOutcome.new(issue: issue), request.accept, 400)
+      end
+
+      def valueset_validates_code(parameters, id_param = nil)
+        # Get a valueset, in one of the many ways a valueset can be specified
+        coding = normalize_valueset_coding(parameters)
+        valueset_response = get_valueset(id_param, find_param(parameters, 'url'))
+        return respond_with_type(valueset_response, request.accept, 400) if valueset_response.is_a? FHIR::OperationOutcome
+
+        validation_fn = FHIR::StructureDefinition.vs_validators[valueset_response]
+
+        if validation_fn
+          code_valid = validation_fn.call(coding)
+          return_params = if code_valid
+                            FHIR::Parameters.new(parameter: [FHIR::Parameters::Parameter.new(name: 'result', valueBoolean: true)])
+                          else
+                            message = "The code '#{coding['code']}' from the code system '#{coding['system']}' is not valid in the valueset '#{valueset_response}'"
+                            params = [
+                              FHIR::Parameters::Parameter.new(name: 'result', valueBoolean: false),
+                              FHIR::Parameters::Parameter.new(name: 'cause', valueString: 'invalid'),
```

Comment:
There is no "cause" parameter defined in $validate-code operation

Reply:
True. Removing that now.